### PR TITLE
Adding Phi-3 model

### PR DESCRIPTION
**Pull request by monk1337**
Adding Phi-3 model

`fsdp_transformer_layer_cls_to_wrap: Phi3DecoderLayer`

### CODE REVIEW

File: examples/phi/phi3-ft.yml [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+  fsdp_limit_all_gathers: true
+  fsdp_sync_module_states: true
+  fsdp_offload_params: true
+  fsdp_use_orig_params: false
+  fsdp_cpu_ram_efficient_loading: true
+  fsdp_auto_wrap_policy: TRANSFORMER_BASED_WRAP
+  fsdp_transformer_layer_cls_to_wrap: Phi3DecoderLayer
+  fsdp_state_dict_type: FULL_STATE_DICT
+  fsdp_sharding_strategy: FULL_SHARD
+resize_token_embeddings_to_32x: true
```
Comment (winglian) [inline examples/phi/phi3-ft.yml]: generally we prefer not to have FSDP or deepspeed specific configs (which require multi-gpu) for smaller models that should fit on single GPU systems.

**Comment by maziyarpanahi** [comment]
Will you also add support for `Phi-3` chat template for fine-tuning? As a reference: https://github.com/unslothai/unsloth/blob/4211cc01409e3ced4f7abebaf68e244193b46e2c/unsloth/chat_templates.py#L269C3-L269C8

**Comment by winglian** [comment]
will merge this after #1582

**Comment by vinamrabenara** [comment]
Was wondering when would this be merged?

**Comment by monk1337** [comment]
> Was wondering when would this be merged?

Too busy to work on this :P

**Comment by maziyarpanahi** [comment]
@winglian I am having some issues with phi-3-small and phi-3-medium models. 

- phi-3-medium with 4k instruct, it constantly fails with `RuntimeError: Cannot re-initialize CUDA in forked subprocess. To use CUDA with multiprocessing, you must use the 'spawn' start method`
- phi-3-small with 8k instruct, fails with:
```
ValueError: For now, we do not support unknown special tokens
In the future, if there is a need for this, we can add special tokens to the tokenizer
starting from rank 100261 - 100263 and then 100266 - 100275.
And finally, we can re-construct the enc object back
```

Is it correct to assume this PR won't solve these, right?

**Comment by winglian** [comment]
@maziyarpanahi yeah, I don't think this PR will resolve that issue.

**Comment by winglian** [comment]
thanks @monk1337 !

**Comment by maziyarpanahi** [comment]
Thank you @winglian and @monk1337

**Pull request merged**